### ISSUE TITLE
Task assignment

### ISSUE THREAD

**Issue by jshorland** [ushahidi/platform#1382]
- [ ] As a user with permissions to manage tasks, I want to assign tasks to a specific user
- [ ] As a user with permissions to manage tasks, I want to un-assign tasks from a specific user
- [ ] As a user with permissions to manage tasks, I want to re-assign tasks from one user to another user
- [ ] As any user, I want to start working on an unassigned task and have it automatically assigned to me
- [ ] As a user assigned to a task, I want to be the only one able to work on the task as long as I am assigned to it
- [ ] As an admin user, I want to assign tasks to specific user roles. i.e have verification be visible ONLY to the verification team etc.

Aha! Link: https://ushahiditeam.aha.io/features/PROD-671

**Comment by brandonrosage** [comment]
An individual post's tasks are dictated by its survey's task configuration. Those tasks are appended to the post automatically, using the structure defined by the survey. That definition includes custom-built tasks, the order of those tasks, whether those tasks require another task be completed first, whether to display the contents of the task with the rest of the post, and who the task is assigned to by default. Furthermore, each task can have custom fields appended to it -- much like survey fields.

### Task creation
![Add task](https://cloud.githubusercontent.com/assets/1136279/20285231/57125e88-aa86-11e6-834c-b331c165b1e8.png)

### Task editing
![Edit task](https://cloud.githubusercontent.com/assets/1136279/20285241/6780d1f0-aa86-11e6-8ef8-0ec9ed8868ff.png)

### Task settings
![Task settings](https://cloud.githubusercontent.com/assets/1136279/20285487/cac0632e-aa87-11e6-901a-6bb4f08c4cee.png)

Once a post is created in a deployment, its survey's tasks are attached to it. And once they're attached, users with permission can change a post's task assignment and status (complete/incomplete) on the fly. This can be done via each task's "---" actions trigger, accessible in three different views: 

### The postcard displayed in "Tasks" mode 
![Postcard](https://cloud.githubusercontent.com/assets/1136279/20285540/2464d8ec-aa88-11e6-9381-016517188771.png)

### The post's detail view 
![post detail](https://cloud.githubusercontent.com/assets/1136279/20285582/56edfe7e-aa88-11e6-8154-3a0b1323c4e3.png)

### The post's edit view
![post edit](https://cloud.githubusercontent.com/assets/1136279/20285605/7994e406-aa88-11e6-8b90-ddd9833bc226.png)

**Comment by brandonrosage** [comment]
@jshorland + @caharding + @rjmackay: This is one of four issues that combine to define the next iteration of "Tasks" for Ushahidi. Those four issues include #1381, #1382, #1383, and #1385.

Though they may not be actively developed in the near term, I'd really like to jump on iterating and testing many of the decisions and assumptions here. So your timely feedback and questions are appreciated. Thanks.

**Comment by jshorland** [comment]
To test: order of completion requirements. I think this is one of the most complex settings

**Comment by rjmackay** [comment]
- How does this interact with publishing/archiving a post? Do users still have the ability make a task required before publishing? Does publishing become a task?
- Should there be a default setting for 'Due date' ie. 1 week after post creation

**Comment by aoduor** [comment]
Adding in some more thoughts onto this.

As an admin user, i'd like to be able to assign tasks to specific user roles. i.e have verification be visible ONLY to the verification team etc.

**Comment by aoduor** [comment]
@jshorland should i create a separate issues for ^^ or is it all good if i just add my notes/user stories in here?

**Comment by brandonrosage** [comment]
@aoduor I consider the user stories defined at the top of the issue (marked as to-do items) as requirements for the design solution. Once I've addressed each of them, my work is done. 

So if you're suggesting new user stories and they're not at the top of an Issue, I can't confidently say I'll address them.

**Comment by aoduor** [comment]
Thanks. I'll add it to the top.

**Comment by aoduor** [comment]
@brandonrosage done :).

**Comment by brandonrosage** [comment]
My revised proposal for the solution to this issue is below, directed by our user interviews in February and the SF Hit Team review April 20. It awaits your review, @jshorland, @natmanning, and @caharding.

A task in Ushahidi is "an explicit indication that a field (or group of fields) needs to be completed; by a specific person or group of people."

You can assign a task to someone (or "anybody") a couple different ways:

1. *On the fly*, for a specific response.
2. *Automatically,* for all incoming responses to a given survey.

## On the fly

When viewing or editing a response, each field and section (a group of fields with its own title, optional description, and permissions) has an "..." action button appended to it.

![field](https://cloud.githubusercontent.com/assets/1136279/25450903/0c50b932-2a86-11e7-9311-e20ded3f6f47.png)

If I select it, I see the option to "Add task."

![add task](https://cloud.githubusercontent.com/assets/1136279/25450950/2f1a0fb8-2a86-11e7-89fb-4bc983bda4aa.png)

And if I select "Add task," a toolbox appears to the right of the field, defining the task that's just been created.

![task toolbox](https://cloud.githubusercontent.com/assets/1136279/25451166/e5f01aca-2a86-11e7-9790-bd24b2172547.png)

I can change the assignee either by selecting the name (which is "anybody," by default) or via the "..." action button appended to the task, which also includes an option to "Remove task."

If I select any part of the screen outside the field or toolbox, the toolbox collapses into just the checkbox.

![collapsed](https://cloud.githubusercontent.com/assets/1136279/25451209/0a879bba-2a87-11e7-898f-6c91d725be01.png)

If I select any part of the field, including any of its inputs, the toolbox expands again. And if I mark the task complete, either by selecting the checkbox or the label, the toolbox takes on a completed state:

![completed](https://cloud.githubusercontent.com/assets/1136279/25451274/3eff27f0-2a87-11e7-9557-d8f717add09a.png)

This completed state is echoed even when it's collapsed:

![completed, collapsed](https://cloud.githubusercontent.com/assets/1136279/25451293/4f69ab2e-2a87-11e7-851b-b3949bb7561c.png)

This allows you to scroll the survey and see at a glance which fields or sections have tasks attached to them and whether they're complete. And to augment that process, there's a standard fixed-positioned toolbox to the top-right of the survey with an interactive summary of the tasks for this response. Selecting them from that list will scroll the window to that field.

![overview](https://cloud.githubusercontent.com/assets/1136279/25451390/a748cf64-2a87-11e7-9786-9cb2e749968c.png)

Note that, in the example above, the entire second section, "Report to department," is assigned to someone.

## Automatically

To configure a survey field so that it's assigned to someone every time a new response is submitted, you'd edit the survey itself, either through Settings > Surveys or the survey filter's action button in the mode context for various modes.

When editing the survey, each of its fields and sections have action buttons attached to them:

![edit survey](https://cloud.githubusercontent.com/assets/1136279/25451547/1c2b6e36-2a88-11e7-9c92-b561332c6c1b.png)

Selecting a field's action button reveals options to "Add task" or "Delete this field." By choosing "Add task," a modal window appears with a summary of the task I'm about to automate, allowing me to confirm what I'm doing, change the assignee, and save:

![add task modal](https://cloud.githubusercontent.com/assets/1136279/25451624/6bbf6c5e-2a88-11e7-8b1e-24425978a0e3.png)

By saving this task, the field's item in the "Edit survey" view will include an indicator that it has a task attached to it (alongside any other applicable indicators):

![field item](https://cloud.githubusercontent.com/assets/1136279/25451688/9ea9d8b6-2a88-11e7-8545-b9985d5e3f39.png)

These indicators are also tooltips which, when selected or moused over, reveal more information about what they mean.

Note that this same convention is adapted for similar actions on sections:

![section toolbar](https://cloud.githubusercontent.com/assets/1136279/25451825/0d3f50f8-2a89-11e7-8879-41fa0f798994.png)

## Configure tasks

Lastly, you see an overview of the tasks that are being automatically assigned in Settings > Tasks. 

![Settings > Tasks](https://cloud.githubusercontent.com/assets/1136279/25451950/7e2f87d8-2a89-11e7-82c7-ddef99de9664.png)

This view is also accessible via the "Configure tasks" button in the toolbar for "Tasks" mode. 

The attached actions (bulk and individual) are the same as those available in comparable views. Selecting a task name in this view would navigate me to its location within its "Edit survey" view. I can change its assignee or remove it faster, however, via the attached action button.
